Avoid unevaluated implicit private

SunilKuravinakop opened the pull request:
For every variable used under `#pragma omp task` directive (`DeclRefExpr`) an ImplicitPrivateVariable is created in the AST, if `private` or `shared` clauses are not present. If the variable has the property of `non_odr_use_unevaluated` e.g. for statements which use `sizeof( i )` `i` will have `non_odr_use_unevaluated` . In such cases CodeGen was asserting by avoiding emitting of LLVM IR for such variables. To prevent this assertion this checkin avoids adding the ImplicitPrivateVariable for variables with `non_odr_use_unevaluated`.

Commit:
Avoiding DeclRefExpr with "non_odr_use_unevaluated" to declare "Implicit Private variable" DeclRefExpr.

  Changes to be committed:
 	modified:   clang/lib/Sema/SemaOpenMP.cpp

Commit:
Adding checks for proper declaration of DeclRefExpr under the task
directive (when variable can be non_odr_use_unevaluated).

  Changes to be committed:
 	modified:   clang/test/OpenMP/task_ast_print.cpp

Commit:
Merge branch 'main' into Avoid_Unevaluated_ImplicitPrivate

Commit:
Merge branch 'main' into Avoid_Unevaluated_ImplicitPrivate

Commit:
Minor changes to the test case.
  Changes to be committed:
 	modified:   clang/test/OpenMP/task_ast_print.cpp

dreachem commented on clang/test/OpenMP/task_ast_print.cpp:
```diff
@@ -208,4 +209,39 @@ int main(int argc, char **argv) {
+    int i = 0;
+    // DUMP: OMPTaskDirective 
+    // DUMP-NEXT: OMPFirstprivateClause
+    // DUMP-NEXT: DeclRefExpr {{.+}} 'i' {{.+}} refers_to_enclosing_variable_or_capture
```
This could still match if `non_odr_use_unevaluated` shows up, can't it? Can we add a check that `non_odr_use_unevaluated` should _not_ be part of the `DeclRefRexpr` under this firstprivate clause?

Commit:
Changes to incorporate Deepak's suggestion:
  This could still match if non_odr_use_unevaluated shows up, can't it?
  Can we add a check that non_odr_use_unevaluated should not be part of
  the DeclRefRexpr under this firstprivate clause?

  An example of DeclRefExpr seen on "-Xclang -ast-dump" is:
     -DeclRefExpr 0x187e0e0 <line:225:6> 'int' lvalue Var 0x187da20 'i' 'int' refers_to_enclosing_variable_or_capture

  Changes to be committed:
 	modified:   clang/test/OpenMP/task_ast_print.cpp

Commit:
Adding DUMP-NOT for check variable 'i':
 DUMP-NOT: DeclRefExpr {{.+}} 'i' {{.+}} non_odr_use_unevaluated

  Changes to be committed:
 	modified:   clang/test/OpenMP/task_ast_print.cpp

alexey-bataev reviewed (approved):
LG

SunilKuravinakop commented:
Alexey,
My colleagues (in my company), having contribution rights, are not around. Can you please merge this change to master?
